cpp/path_module/algorithm/path.cpp

```diff
@@ -304,115 +346,90 @@ void Path::Expand(mgp_list *args, mgp_graph *memgraph_graph, mgp_result *result,
   }
 }
 
-void GetStartNodes(const mgp::Value element, const mgp::Graph &graph, std::unordered_set<mgp::Node> &start_nodes) {
-  if (!(element.IsNode() || element.IsInt())) {
+void Path::PathSubgraph::Parse(const mgp::Value &value) {
+  if (!(value.IsNode() || value.IsInt())) {
     throw mgp::ValueException("The first argument needs to be a node, an integer ID, or a list thereof.");
   }
-  if (element.IsNode()) {
-    start_nodes.insert(element.ValueNode());
+  if (value.IsNode()) {
+    start_nodes_.insert(value.ValueNode());
     return;
   }
-  start_nodes.insert(graph.GetNodeById(mgp::Id::FromInt(element.ValueInt())));
+  start_nodes_.insert(graph_.GetNodeById(mgp::Id::FromInt(value.ValueInt())));
 }
 
-bool RelFilterAllows(const mgp::Map &config, std::string_view type, bool ingoing) {
-  mgp::List list_of_types = config.At("relationshipFilter").ValueList();
-  if (list_of_types.Size() == 0) {
-    return true;
-  }
-  for (const auto element : list_of_types) {
-    auto string_rel_type = element.ValueString();
-    if (string_rel_type.front() == '<' && ingoing) {
-      string_rel_type.remove_prefix(1);
-    }
-    if (string_rel_type.back() == '>' && !ingoing) {
-      string_rel_type.remove_suffix(1);
-    }
-    if (string_rel_type == type || string_rel_type.size() == 0) {
-      return true;
+void Path::PathSubgraph::Expand(Pair &pair, mgp::Relationships relationships, bool outgoing, std::queue<Pair> &queue,
+                                std::set<std::pair<std::string_view, int64_t>> &seen) {
+  for (const auto relationship : relationships) {
+    auto next_node = outgoing ? relationship.To() : relationship.From();
+    auto type = std::string(relationship.Type());
+    auto wanted_direction = helper_.GetDirection(type);
+
+    if (helper_.FilterRelationships(pair.hop_count == 0)) {
+      if ((wanted_direction == RelDirection::kNone && !helper_.AnyDirected(outgoing)) ||
+          visited_.contains(next_node.Id().AsInt())) {
+        continue;
+      }
     }
-  }
-  return false;
-}
 
-bool IsLabelListed(const mgp::Node node, std::unordered_set<std::string_view> &set) {
-  for (const auto label : node.Labels()) {
-    if (set.contains(label)) {
-      return true;
+    RelDirection curr_direction = outgoing ? RelDirection::kOutgoing : RelDirection::kIncoming;
+
+    if (wanted_direction == RelDirection::kAny || curr_direction == wanted_direction || helper_.AnyDirected(outgoing)) {
+      visited_.insert(next_node.Id().AsInt());
+      queue.push({next_node, pair.hop_count + 1});
```

Comment:
This is an optimisation that can sometimes help a lot. If a node is unvisited you know that the current path you're on is the shortest path since you are running BFS so it's always the correct choice. If you don't mark it now you'll just add a lot of nodes in the queue that will be skipped over later because when you get to them the first path that expanded to them will already visit them and render them useless.

Reply:
I can elaborate further on daily if needed.

Reply:
Ah, okay, I think I get it, good